### ISSUE TITLE
ICE: `b` is not bound to a value

### ISSUE THREAD

**Issue by Snektron** [diku-dk/futhark#1538]
While optimizing advent of code 2021, day 3, ran into the following internal compiler error:
```
Internal compiler error (unhandled IO exception).
Please report this at https://github.com/diku-dk/futhark/issues
`b` is not bound to a value.
CallStack (from HasCallStack):
  error, called at src/Language/Futhark/Interpreter.hs:707:10 in futhark-0.21.0-IdT9ZmWZph2HTE2hcJfaJe:Language.Futhark.Interpreter
```

With the following code:
```fut
let parse [n] (bits: i64) (input: [n]u8) =
    input
    |> unflatten (n / (bits + 1)) bits
    |> map (\line -> map2 (\c i -> 0) line (indices line))

entry main (input: []u8) = parse 12 input
```

**Comment by Snektron** [comment]
Should be noted that it works fine with `futhark c` so this is an issue with only the interpreter.

**Comment by Snektron** [comment]
I also forgot to mention that the error happens with this input: <https://github.com/Snektron/aoc21/blob/main/input/03.in>. I use <https://github.com/Snektron/aoc21/blob/main/scripts/stringify.py> to translate the input into a string accepted by `futhark run`:
`cat input/03.in | scripts/stringify.py | futhark run 03.fut -w -e main`.

**Comment by athas** [comment]
I think my fix of #1539 addresses this.  Could you check?  At least with that input I now get a proper error.

**Comment by athas** [comment]
Fixed as part of #1539.
